plugins/common/lambda-rules-registration.hpp

```diff
@@ -0,0 +1,131 @@
+ *
+ * At minimum, the rule string and if_lambda need to be set.
+ */
+struct lambda_rule_registration_t
+{
+    /**
+     * @brief rule This is the rule text.
+     *
+     * @note The registering plugin is supposed to set this value before registering.
+     */
+    std::string rule;
+
+    /**
+     * @brief if_lambda This is the lambda method to be executed if the specified condition holds.
+     *
+     * @note The registering plugin is supposed to set this value before registering.
+     */
+    wf::lambda_t if_lambda;
+
+    /**
+     * @brief else_lambda This is the lambda method to be executed if the specified condition does not hold.
+     *
+     * @note The registering plugin is supposed to set this value before registering.
+     * @note In most cases this should be left blank.
+     *
+     * @attention: Be very careful with this lambda because it will be executed on the signal for each view
+     *             that did NOT match the condition.
+     */
+    wf::lambda_t else_lambda;
+
+    /**
+     * @brief access_interface Access interface to be used when evaluating the rule.
+     *
+     * @note If this is left blank (nullptr), the standard view_access_interface_t instance will be used.
+     */
+    std::shared_ptr<wf::access_interface_t> access_interface;
+
+    /**
+     * @brief rule_instance Pointer to the parsed rule object.
+     *
+     * @attention You should not set this. Leave it at nullptr, the registration process will fill in this
+     *            variable. Window rules can then use this cached rule instance on each signal occurrence.
+     */
+    std::shared_ptr<wf::lambda_rule_t> rule_instance;
+};
+
+/**
+ * @brief The lambda_rules_registrations_t class is a helper class for easy registration and unregistration of
+ *        lambda rules for the window rules plugin.
+ *
+ * This class is a singleton and can only be used via the getInstance() method.
+ *
+ * The instance is stored in wf::core. The getInstance() method will fetch from wf:core, and lazy-init if the
+ * instance is not yet present.
+ */
+class lambda_rules_registrations_t : public custom_data_t
+{
+  public:
+    /**
+     * @brief getInstance Static accessor for the singleton.
+     *
+     * @return Observer pointer to the singleton instance, fetched from wf::core.
+     */
+    static nonstd::observer_ptr<lambda_rules_registrations_t> getInstance();
+
+    /**
+     * @brief registerLambdaRule Registers a lambda rule with its associated key.
+     *
+     * This method will return error result if the key is not unique or the registration struct is incomplete.
+     *
+     * @param[in] key Unique key for the registration.
+     * @param[in] registration The registration structure.
+     *
+     * @return <code>True</code> in case of error, <code>false</code> if ok.
+     */
+    bool registerLambdaRule(std::string key, std::shared_ptr<lambda_rule_registration_t> registration);
```

Comment:
Trying to clarify things for myself: why do we use a `shared_ptr` here? Who is the owner of the memory? Isn't this always the plugin using this interface? (so it could be observer_ptr, if I understand correctly).

Reply:
I used shared_ptr here because the registration object will be changed (compiled rule added). I will read up on observer_ptr to see if I can use that instead. This whole smart pointer stuff is not something I use a lot (Qt and embedded dev) and I am bound to fuck up here and there.

Reply:
Hmm, in that case, maybe you want to pass a unique_ptr ? This way the plugin calling this transfers the object to window-rules and does not own it anymore.

Reply:
I guess a shared_ptr is fine enough.